keps/sig-node/0000-apparmor-ga/README.md

```diff
@@ -0,0 +1,685 @@
+always create _new profiles_ instead of updating existing ones. Create and
+deploy a new version of the existing Pod Template, changing the profile name to
+the newly created profile. Redeploy, once working delete the former Pod
+Template. This will avoid disruption on in-flight workloads.
+
+The current behavior lacks features to facilitate the maintenance of AppArmor
+profiles across the cluster. Two examples being: 1) the lack of profile
+synchronization across nodes and 2) how difficult it can be to identify that
+profiles have been changed on disk, after pods started using it. However, given
+its current "pseudo-GA" state, we don't want to change it with this KEP.
+
+###### Profile files managed by the cluster admins
+
+The current support relies on profiles being saved as files in all cluster nodes
+where the pods using them may be scheduled. It is also the cluster admin's
+responsibility to ensure the profiles are correctly saved and synchronised
+across the all nodes.
+
+##### RuntimeDefault Profile
+
+We propose maintaining the support to a single runtime profile, which will be
+defined by using the `AppArmorProfileTypeRuntimeDefault`. The reasons being:
+
+- No changes to the current behavior. Users are currently not allowed to specify
+  other runtime profiles. The existing API server rejects runtime profile names
+  that are different than `runtime/default`. The only exception being
+  `docker/default` - for backwards compatibility.
+- Most runtimes only support the default profile, although the CRI is flexible
+  enough to allow the kubelet to send other profile names.
+- Dockershim does not currently provide a way to pass other runtime profile
+  names.
+- Multiple runtime profiles has never been requested as a feature.
+
+If built-in support for multiple runtime profiles is needed in the future, a new
+KEP will be created to cover its details.
+
+## Design Details
+
+### Failure and Fallback Strategy
+
+There are different scenarios in which applying an AppArmor profile may fail,
+below are the ones we mapped and their outcome once this KEP is implemented:
+
+| Scenario                                                                                           | API Server Result    | Kubelet Result                                                                                                                                        |
+| -------------------------------------------------------------------------------------------------- | -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 1) Using custom or `runtime/default` profile when container runtime does not support AppArmor.     | Pod created          | The outcome is container runtime dependent. In this scenario containers may 1) fail to start or 2) run normally without having its policies enforced. |
```

Comment:
There's another case to list here:

Host kernel or build does not have apparmor enabled - Kubelet will "soft fail" the pod by putting it in the `Blocked` state.

Reply:
Added :+1: